### PR TITLE
Automatic fw upgrade for mlnx platform (#18)

### PR DESCRIPTION
**Pull request by marian-pritsak**
Implement script for firmware upgrade to required version
Add firmware binary and script to ops-syncd-mlnx container
Add pciutils and usbutils to sonic-generic.bin

Signed-off-by: marian-pritsak <marianp@mellanox.com>

### CODE REVIEW

**Comment by qiluo-msft** [comment]
Why including firmware upgrade into syncd docker? Is it one-time thing?

File: build_debian.sh [hunk 1]
```diff
@@ -183,7 +183,9 @@ sudo LANG=C chroot $FILESYSTEM_ROOT apt-get -y install      \
     bsdmainutils            \
     ca-certificates         \
     i2c-tools               \
-    efibootmgr
+    efibootmgr              \
+    usbutils                \
```
Comment (lguohan) [inline build_debian.sh]: which component requires this usbutils?

File: dockers/docker-syncd-mlnx/start.sh [hunk 2]
```diff
@@ -8,6 +8,10 @@ function clean_up {
 
 trap clean_up SIGTERM SIGKILL
 
+# fw-upgrade will exit if firmware was actually upgraded or if some error
+# occures
+. mlnx-fw-upgrade.sh
```
Comment (lguohan) [inline dockers/docker-syncd-mlnx/start.sh]: after upgrading the fw, do we need to restart the sxd kernel driver?

Comment (marian-pritsak) [inline dockers/docker-syncd-mlnx/start.sh]: Yes, we need to restart sxd kernel in order to run newly burned firmware

File: build_debian.sh (hunk 1 above)
Comment (marian-pritsak) [inline build_debian.sh]: mst utility uses usbutils and pciutils to talk to ASIC

**Comment by marian-pritsak** [comment]
As MLNX SDK is shipped with syncd docker, firmware upgrade utility is also going to be a part of it, so that in case of rolling out a new version of SDK, a compatible version of firmware will be installed.

During syncd start firmware version is verified to be compatible with SDK shipped with current syncd, and if not - upgrade procedure is initiated. Its a one time action, which happens only when new SDK is available in syncd.

File: Makefile [hunk 3]
```diff
@@ -73,6 +73,8 @@ $(addprefix dockers/docker-syncd-mlnx/deps/,syncd_1.0.0_amd64.deb libsairedis_1.
 	mkdir -p `dirname $@` && cp $< $(dir $@)
 dockers/docker-syncd-mlnx/deps/%.deb: src/%.deb
 	mkdir -p `dirname $@` && cp $< $(dir $@)
+dockers/docker-syncd-mlnx/deps/fw-SPC.mfa: src/mlnx-sdk/fw-SPC.mfa
```
Comment (stcheng) [inline Makefile]: where this folder mlnx-sdk is created?

File: dockers/docker-syncd-mlnx/start.sh (hunk 2 above)
Comment (marian-pritsak) [inline dockers/docker-syncd-mlnx/start.sh]: Yes, correct
If firmware is actually updated, mlnx-fw-upgrade.sh will exit, causing syncd service to restart

Comment (qiluo-msft) [inline dockers/docker-syncd-mlnx/start.sh]: It is actually a trick to exit directly from mlnx-fw-upgrade.sh. Please add your comment as code comment.

BTW, syncd service to restart ==  restart sxdkernel driver?

File: dockers/docker-syncd-mlnx/mlnx-fw-upgrade.sh [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+if [[ -z ${fw_version} ]]; then
+	echo "Could not retreive current FW version."
+	exit 1
+fi
+
+if [[ ${required_fw_version} == ${fw_version} ]]; then
+	echo "Mellanox firmware is up to date."
+else
+	echo "Mellanox firmware required version is ${required_fw_version}. Upgrading..."
```
Comment (qiluo-msft) [inline dockers/docker-syncd-mlnx/mlnx-fw-upgrade.sh]: Downgrading is also possible here. Is it safe to downgrade? Maybe you can refine the message.

File: dockers/docker-syncd-mlnx/start.sh (hunk 2 above)
Comment (stcheng) [inline dockers/docker-syncd-mlnx/start.sh]: @qiluo-msft yes. restart syncd service will restart sxdkernel driver

Comment (marian-pritsak) [inline dockers/docker-syncd-mlnx/start.sh]: > It is actually a trick to exit directly from mlnx-fw-upgrade.sh. Please add your comment as code comment.

There is a comment in start.sh line 11 and mlnx-fw-upgrade.sh line 51 regarding restart logic. Anything else needed?

Comment (qiluo-msft) [inline dockers/docker-syncd-mlnx/start.sh]: No, thanks!

**Pull request merged**